Convert to OleDb assembly and last Cryptography DllImport to GeneratedDllImport

AaronRobinsonMSFT opened the pull request:
/cc @jkoritzinsky @elinor-fung

AaronRobinsonMSFT committed:
Enable DllImport generator for all source projects.

AaronRobinsonMSFT committed:
Convert OleDb to only use GeneratedDllImport.

AaronRobinsonMSFT committed:
Remove last DllImport from Cryptography assemblies.

jkoritzinsky reviewed (commented):
While you're here, can you update the OleDb P/Invokes to match the interop guidelines for the repo? (putting them in the Interop/Windows folder, etc?)

AaronRobinsonMSFT committed:
Update source generator to distinguish between "no marshalling info"
  which can cause a failure vs "missing support" which enables us to
  fallback to the marshalling forwarder.

AaronRobinsonMSFT committed:
Remove unused DllImports.

AaronRobinsonMSFT commented:
> While you're here, can you update the OleDb P/Invokes to match the interop guidelines for the repo? (putting them in the Interop/Windows folder, etc?)

I don't think I can do that in this review to be honest. The signatures don't all match and uses of those types aren't the same in all places. This would mean I'd need to update many other projects. I'm not sure how we should be doing this since the model in place now means we need to trace through all projects and permit generation which makes many of these changes very hard to reason about.

AaronRobinsonMSFT committed:
Remove local CoTaskMemAlloc and CoTaskMemFree.

AaronRobinsonMSFT committed:
Reuse existing P/Invoke definitions.

AaronRobinsonMSFT committed:
Replace local SysAllocStringLen with a Common define.

AaronRobinsonMSFT committed:
Fix SysAllocStringLen.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into convert_to_generateddllimport

AaronRobinsonMSFT committed:
Update Crypto P/Invokes.

elinor-fung commented on src/libraries/Common/src/Interop/Windows/OleAut32/Interop.SetLastError.cs:
```diff
@@ -0,0 +1,16 @@
+        // only using this to clear existing error info with null
+        [GeneratedDllImport(Interop.Libraries.OleAut32)]
+        // TLS values are preserved between threads, need to check that we use this API to clear the error state only.
+        internal static partial void SetErrorInfo(int dwReserved, IntPtr pIErrorInfo);
```
File should be `Interop.SetErrorInfo.cs` instead of `Interop.SetLastError.cs`

elinor-fung commented on src/libraries/Common/src/Interop/Windows/Ole32/Interop.PropClearVariant.cs:
```diff
@@ -0,0 +1,14 @@
+    internal static partial class Ole32
+    {
+        [GeneratedDllImport(Interop.Libraries.Ole32)]
+        internal static partial void PropVariantClear(IntPtr pObject);
```
`Interop.PropClearVariant.cs` -> `Interop.PropVariantClear.cs`

elinor-fung commented on eng/generators.targets:
```diff
@@ -35,8 +41,8 @@
     <Compile Include="$(LibrariesProjectRoot)Common\src\System\Runtime\InteropServices\GeneratedDllImportAttribute.cs" />
-    <Compile Include="$(LibrariesProjectRoot)Common\src\System\Runtime\InteropServices\GeneratedMarshallingAttribute.cs" />
-    <Compile Include="$(LibrariesProjectRoot)Common\src\System\Runtime\InteropServices\ArrayMarshaller.cs" />
+    <Compile Condition="'$(NetCoreAppCurrentTargetFrameworkMoniker)' == '$(TargetFrameworkMoniker)'" Include="$(LibrariesProjectRoot)Common\src\System\Runtime\InteropServices\GeneratedMarshallingAttribute.cs" />
```
Can we add a comment why we do this conditional include?

elinor-fung commented on src/libraries/Common/src/Interop/Windows/Kernel32/Interop.LocalAlloc.cs:
```diff
@@ -10,6 +10,7 @@ internal static partial class Kernel32
         internal const uint LMEM_MOVEABLE = 0x0002;
+        internal const uint LMEM_ZEROINIT = 0x0040;
 
         [DllImport(Libraries.Kernel32)]
```
Should we just make these `GeneratedDllImport` too? Since we do the extern partial forwarder for blittable functions now.

AaronRobinsonMSFT replied:
They are used in other projects so I didn't update them. I tried to only update P/Invokes that I knew worked everywhere. Some consumers don't have the needed references.

AaronRobinsonMSFT commented on src/libraries/Common/src/Interop/Windows/Crypt32/Interop.CryptMsgGetParam.cs:
```diff
@@ -11,51 +11,44 @@ internal static partial class Crypt32
+            out CryptMsgType pvData,
+            ref int pcbData);
 
         [DllImport(Libraries.Crypt32, CharSet = CharSet.Unicode, SetLastError = true)]
```
Sure.

AaronRobinsonMSFT committed:
Review feedback.

jkoritzinsky commented on eng/generators.targets:
```diff
@@ -27,6 +27,12 @@
                             and @(Reference->AnyHaveMetadataValue('Identity', 'System.Memory')))
                           or ('@(ProjectReference)' != ''
                             and @(ProjectReference->AnyHaveMetadataValue('Identity', '$(CoreLibProject)'))))" />
+    <EnabledGenerators Include="DllImportGenerator"
+                       Condition="'$(EnableDllImportGenerator)' == ''
+                        and '$(IsFrameworkSupportFacade)' != 'true'
+                        and '$(IsSourceProject)' == 'true'
+                        and '$(MSBuildProjectExtension)' == '.csproj'
+                        and '$(TargetFrameworkIdentifier)' == '.NETStandard'" />
```
Do we need to add a .NET Framework case as well? I know we discussed this a little offline, but I don't remember what conclusion we came to.

AaronRobinsonMSFT replied:
I don't think so. I think this works as is, but will monitor it.
